harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -116,6 +204,135 @@ def _set_random_seeds(cls: Type["PyTorchTrialController"], seed: int) -> None:
+                self.searcher_op.report_completed(searcher_metric)
+
+        self.core_context.train.report_validation_metrics(steps_completed, metrics)
+
+        if self.state.last_ckpt != self.steps_completed:
+            if self.ckpt_policy == "all" or (
+                self.ckpt_policy == "best"
+            ):
+                self.checkpoint(already_exiting=False)
+            elif self.ckpt_policy == "best" and searcher_metric:
+                self.is_best_validation(now=searcher_metric, before=self.core_context.train.get_experiment_best_validation())
+                self.checkpoint(already_exiting=False)
+
+    def is_best_validation(self, now: float, before: Optional[float]) -> bool:
+        if before is None:
+            return True
+
+        return (now < before) if self.smaller_is_better else (now > before)
+
+    def on_epoch_start(self, epoch_idx: int):
+        for callback in self.callbacks.values():
+            with self.prof.record_timing(
+                f"callbacks.{callback.__class__.__name__}.on_training_epoch_start"
+            ):
+                sig = signature(callback.on_training_epoch_start)
+                if sig.parameters:
+                    callback.on_training_epoch_start(epoch_idx)
+                else:
+                    logging.warning(
+                        "on_training_epoch_start() without parameters is deprecated"
+                        " since 0.17.8. Please add epoch_idx parameter."
+                    )
+                    callback.on_training_epoch_start()  # type: ignore[call-arg]
+
+    def on_epoch_end(self, epoch_idx: int):
+        for callback in self.callbacks.values():
+            with self.prof.record_timing(
+                f"callbacks.{callback.__class__.__name__}.on_training_epoch_end"
+            ):
+                callback.on_training_epoch_end(epoch_idx)
+
+    def checkpoint(self, already_exiting: bool):
+        if self.state.last_ckpt == self.steps_completed:
+            return
+
+        self.core_context.train.set_status("checkpointing")
+        self.state.last_ckpt = self.steps_completed
+
+        uuid = ""
+        try:
+            if self.is_chief:
+                metadata = {
+                    "determined_version": det.__version__,
+                    "steps_completed": self.steps_completed,
+                    "framework": f"torch-{torch.__version__}",
+                    "format": "pickle",
+                }
+                with self.context._core.checkpoint.store_path(metadata) as (
+                    path,
+                    storage_id,
+                ):
+                    self._save(path)
+                    uuid = storage_id
+            uuid = self.context.distributed.broadcast(uuid)
+            for callback in self.callbacks.values():
+                callback.on_checkpoint_upload_end(uuid=uuid)
+        except det.InvalidHP:
+            self.core_context.train.report_early_exit(core.EarlyExitReason.INVALID_HP)
```

Comment:
This is already done inside the `with det.core.init()` call, right?  I don't think it's needed here.

Reply:
talked about this.  It may be necessary to watch for the exception and save a checkpoint or something (maybe?) but we should not be reporting early exit twice I think.

Reply:
i think this is needed here because the checkpoint workload needs to know whether to skip exit checkpoint. so we catch and re-raise all other InvalidHP outside of checkpoint workloads for handling by the upstream core API

Reply:
I see, but let's only report_early_exit when we have to raise the ShouldExit, and just re-raise the other times.